### PR TITLE
cram: add package for building cram support on

### PR DESCRIPTION
**Pull request by kortschak**
@brentp I think we should merge this as a basis for continuing work. It's not going to attract  work in its own branch.

I will add docs and have a read through over the next couple of days, so I won't need an approval until then.

### CODE REVIEW

**Comment by kortschak** [comment]
PTAL @brentp 

Also @sb10, please feel free to have a look.

File: cram/cram.go [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+}
+
+// readFrom populates a definition from the given io.Reader. If the magic
+// number of the file is not "CRAM" readFrom returns an error.
+func (d *definition) readFrom(r io.Reader) error {
+	err := binary.Read(r, binary.LittleEndian, d)
+	if err != nil {
+		return err
+	}
+	magic := reflect.TypeOf(*d).Field(0).Tag.Get("is")
```
Comment (brentp) [inline cram/cram.go]: this seems like a complicated way to get `[]byte("CRAM")`

File: cram/cram.go [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+	case bzip2Method:
+		return ioutil.ReadAll(bzip2.NewReader(bytes.NewReader(b.blockData)))
+	case lzmaMethod:
+		lz, err := lzma.NewReader(bytes.NewReader(b.blockData))
+		if err != nil {
+			return nil, err
+		}
+		return ioutil.ReadAll(lz)
+	case ransMethod:
+		// Unimplemented.
```
Comment (brentp) [inline cram/cram.go]: this could panic or return error

File: cram/encoding/ltf8/ltf.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	5:  6,
+	6:  1,
+	9:  4,
+	10: 2,
+	11: 5,
+	14: 0,
+}
+
+// nlo returns the number of leading set bits in x.
+func nlo(x byte) int {
```
Comment (brentp) [inline cram/encoding/ltf8/ltf.go]: `nlo(x)` (and pop) can be replace by `bits.LeadingZeros8(^x)`

File: cram/encoding/itf8/itf.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	5:  6,
+	6:  1,
+	9:  4,
+	10: 2,
+	11: 5,
+	14: 0,
+}
+
+// nlo returns the number of leading set bits in x.
+func nlo(x byte) int {
```
Comment (brentp) [inline cram/encoding/itf8/itf.go]: repalce with `bits.LeadingZeros8(^x)`

File: cram/encoding/ltf8/ltf.go (hunk 3 above)
Comment (brentp) [reply cram/encoding/ltf8/ltf.go]: I guess `bits` dependency would truncate support to more recent versions of go.

Comment (kortschak) [reply cram/encoding/ltf8/ltf.go]: I'm happy to drop that. "math/bits" didn't exist when I wrote this.

File: cram/cram.go (hunk 1 above)
Comment (kortschak) [reply cram/cram.go]: At the time, it helped me keep the ownership details under control. If it's a performance problem we can revisit.

File: cram/encoding/itf8/itf.go (hunk 4 above)
Comment (kortschak) [reply cram/encoding/itf8/itf.go]: Done

File: cram/encoding/ltf8/ltf.go (hunk 3 above)
Comment (kortschak) [reply cram/encoding/ltf8/ltf.go]: Done

File: cram/cram.go (hunk 2 above)
Comment (kortschak) [reply cram/cram.go]: Done.

**Pull request merged**